### PR TITLE
 Add Adsbrain endpoint [Official PR]

### PR DESCRIPTION
**Pull request by JiushengChen**

### CODE REVIEW

File: src/http_server.cc [hunk 1]
```diff
@@ -2173,6 +2173,12 @@ HTTPAPIServer::EVBufferToInput(
     }
   }
 
+  // get real header length
+  if (n > 0 && typeid(*this) != typeid(HTTPAPIServer)) {  // only for adsbrain
+    header_length = *(unsigned int *)((char *)v[n-1].iov_base + v[n-1].iov_len - 4);
```
Comment (feihugis) [inline src/http_server.cc:2179]: What does `4` refer to here?

File: src/http_server.h [hunk 2]
```diff
@@ -221,7 +221,7 @@ class HTTPAPIServer : public HTTPServer {
     std::atomic<uint32_t> response_count_;
   };
 
- protected:
+ public:
```
Comment (feihugis) [inline src/http_server.h]: Curios why need to change it to `public`?

File: src/main.cc [hunk 3]
```diff
@@ -87,7 +91,7 @@ int32_t repository_poll_secs_ = 15;
 // default values and modifyied based on command-line args.
 #ifdef TRITON_ENABLE_HTTP
 std::unique_ptr<triton::server::HTTPServer> http_service_;
-bool allow_http_ = true;
+bool allow_http_ = false;
```
Comment (feihugis) [inline src/main.cc:94]: Is this change needed?

File: src/main.cc [hunk 4]
```diff
@@ -117,7 +137,7 @@ std::string vertex_ai_default_model_;
 
 #ifdef TRITON_ENABLE_GRPC
 std::unique_ptr<triton::server::GRPCServer> grpc_service_;
-bool allow_grpc_ = true;
+bool allow_grpc_ = false;
```
Comment (feihugis) [inline src/main.cc:140]: Is this change needed?

File: src/main.cc [hunk 5]
```diff
@@ -1690,6 +1794,60 @@ Parse(TRITONSERVER_ServerOptions** server_options, int argc, char** argv)
   sagemaker_safe_range_ = sagemaker_safe_range;
 #endif  // TRITON_ENABLE_SAGEMAKER
 
+#if defined(TRITON_ENABLE_ADSBRAIN)
+  adsbrain_port_ = adsbrain_port;
+  adsbrain_thread_cnt_ = adsbrain_thread_cnt;
+
+  // entrypoint
+  char* v = getenv(adsbrain_entrypoint_env.c_str());
```
Comment (feihugis) [inline src/main.cc:1802]: A doc about how to run triton-adsbrain-server will be helpful.

File: src/http_server.cc (hunk 1 above)
Comment (JiushengChen) [reply src/http_server.cc:2179]: sizeof(unsigned int), the additional value appended.

File: src/main.cc (hunk 3 above)
Comment (JiushengChen) [reply src/main.cc:94]: http and grpc are started by default. They are not needed in our case, therefore disable to save some system resources.

File: src/main.cc (hunk 4 above)
Comment (JiushengChen) [reply src/main.cc:140]: same as above

File: src/http_server.h (hunk 2 above)
Comment (JiushengChen) [reply src/http_server.h]: unnecessary. will roll back.

File: src/main.cc (hunk 5 above)
Comment (JiushengChen) [reply src/main.cc:1802]: will update our docs in sharepoint.

File: src/http_server.cc [hunk 6]
```diff
@@ -2173,6 +2173,19 @@ HTTPAPIServer::EVBufferToInput(
     }
   }
 
+  // get real header length, only for adsbrain
+  if (n > 0 && typeid(*this) != typeid(HTTPAPIServer)) {
+    header_length = *(unsigned int *)((char *)v[n-1].iov_base + v[n-1].iov_len - 4);
```
Comment (feihugis) [inline src/http_server.cc]: Could you explain the logic a little bit here to help me understand the code? I'm not clear about what `iov_base` and `iov_len` mean here.

For Line 2177 `typeid(*this) != typeid(HTTPAPIServer)`, would it be better to be something like `typeid(*this) == AdsbrainServerID`? There may be other extended servers such as Sagemaker, but the code section is only for Adsbrain server.

Comment (feihugis) [reply src/http_server.cc]: What will the request body look like? Just curious how the request body from DLIS is handled here.

Comment (JiushengChen) [reply src/http_server.cc]: According to c++ doc, typeid output is not fixed, can vary. So a fixed value there may not work someday later. 
The body is constructed by our client, dlis just passes on the data. The body format is:
"\<json\>+\<tensor\>+\<lengthOfJson\>"

now.

Comment (feihugis) [reply src/http_server.cc]: It should be fine to our case. But it actually may break other servers. We can leave a TODO here. One cleaner solution may be: add the missing header length info to `req` inside the  function of `AdsBrainAPIServer::Handle(evhtp_request_t* req)`. Not sure if it is doable or any performance overhead.

File: docs/adsbrain.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      "name" : "output0",
+      "parameters" : {
+        "binary_data" : true
+      }
+    }
+  ]
+}
+<16 bytes of data for input0 tensor>
+<3 bytes of data for input1 tensor>
+<4 bytes for jason part length>
```
Comment (feihugis) [inline docs/adsbrain.md]: Typo: `jason` -> `json`.

File: src/http_server.cc (hunk 6 above)
Comment (JiushengChen) [reply src/http_server.cc]: it won't break other servers because of line 2177. The other option of modifying req looks heavy because it involves memory allocation and thread lock.

File: src/main.cc (hunk 5 above)
Comment (JiushengChen) [reply src/main.cc:1802]: added doc.

File: src/http_server.cc (hunk 6 above)
Comment (feihugis) [reply src/http_server.cc]: Will Line 2177 be true  for any other extended server (e.g., sagemaker_server)?

Comment (JiushengChen) [reply src/http_server.cc]: fixed

File: src/main.cc [hunk 8]
```diff
@@ -847,6 +910,18 @@ StartEndpoints(
   }
 #endif  // TRITON_ENABLE_SAGEMAKER
 
+#ifdef TRITON_ENABLE_ADSBRAIN
+  // Enable Sagemaker endpoints if requested...
```
Comment (feihugis) [inline src/main.cc]: typo: `Sagemaker` -> `Adsbrain`

**Review by feihugis — APPROVED** [review]
LGTM!

File: src/http_server.cc [hunk 9]
```diff
@@ -29,6 +29,7 @@
 #endif
 
 #include "http_server.h"
+#include "adsbrain_server.h"
```
Comment (feihugis) [inline src/http_server.cc:32]: This seems to cause circular dependency. Did you test if the code works after this change?

File: src/main.cc (hunk 8 above)
Comment (JiushengChen) [reply src/main.cc]: will fix

File: src/http_server.cc (hunk 9 above)
Comment (JiushengChen) [reply src/http_server.cc:32]: this is .cc file. no problem. Tested.
